projects/convai2/baselines/profilememory/eval_ppl.py

```diff
@@ -0,0 +1,166 @@
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+from parlai.core.agents import create_agent, create_agents_from_shared
+from parlai.core.build_data import download_models
+from projects.convai2.build_dict import build_dict
+from projects.convai2.eval_ppl import setup_args, eval_ppl
+from parlai.core.dict import DictionaryAgent
+from parlai.core.params import ParlaiParser
+from parlai.core.thread_utils import SharedTable
+from parlai.core.utils import Timer, round_sigfigs, no_lock
+from parlai.core.worlds import World, create_task
+from projects.personachat.persona_seq2seq import PersonachatSeqseqAgentSplit
+import torch.nn.functional as F
+from torch.autograd import Variable
+from torch.autograd import Variable
+import torch
+import math
+import random
+
+
+class ProfileMemoryEntry(PersonachatSeqseqAgentSplit):
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        if shared:
+            self.probs = shared['probs']
+        else:
+            # default minimum probability mass for all tokens
+            self.probs = {k: 1e-7 for k in build_dict().keys()}
+
+    def share(self):
+        shared = super().share()
+        shared['probs'] = self.probs.copy()
+        return shared
+
+    def next_word_probability(self, observation, partial_out):
+        """Return probability distribution over next words given an input and
+        partial true output. This is used to calculate the per-word perplexity.
+        Arguments:
+        observation -- input observation dict
+        partial_out -- previous "true" words
+        Returns a dict, where each key is a word and each value is a probability
+        score for that word. Unset keys assume a probability of zero.
+        e.g.
+        {'text': 'Run test program.'}, ['hello'] => {'world': 1.0}
+        """
+        if not hasattr(self, 'last_text'):
+            self.last_text = None
+
+        if observation['text'] != self.last_text:
+            self.last_text = observation.get('text')
+            self.observe(observation)
```

Comment:
can remove all the `self.last_text` stuff, not being used below

Reply:
it is used to check whether or not we already observed the text

Reply:
ahhh yes I see I see